### DISCUSSION TITLE
Why `switch_session()` doesn't work?

### DISCUSSION THREAD

**Discussion by Andrew15-5** [zellij-org/zellij#3329] [Q&A] [answered]
I created the basic plugin:

```rs
use zellij_tile::prelude::*;
use std::collections::BTreeMap;

#[derive(Default)]
struct State {}

register_plugin!(State);

impl ZellijPlugin for State {
    fn load(&mut self, _configuration: BTreeMap<String, String>) {
        switch_session(Some("existing-session"));
        close_self();
    }
    fn update(&mut self, _event: Event) -> bool {
        false
    }
    fn render(&mut self, _rows: usize, _cols: usize) {}
}
```

But it doesn't actually switch to the existing session — it only closes itself. I tried putting it into the other two callbacks, but it didn't change anything. I'm trying to achieve what is described in https://github.com/zellij-org/zellij/discussions/2448#discussioncomment-7803560. `zj a session-name` doesn't work anymore because I now already start the shell inside the zellij. I need to remove the nestingness. I need an instant non-interactive result just like `zj a <session>`. How can I achieve this?

My only current "solution" is to start a new terminal with the desired zellij session (using an alias). The other "solution" would be to not run shell from the zellij by default, but that partially defeats the purpose of the multiplexer.

**Comment by cristiand391** [comment]
the `switch_session` command requires the `ChangeApplicationState` permission:
https://zellij.dev/documentation/plugin-api-commands#switch_session
https://zellij.dev/documentation/plugin-api-permissions

you need to request it in the `load` function like this:
https://github.com/cristiand391/zj-status-bar/blob/5048c3c841d47b2d35def393bf0f5676ed68e42d/src/main.rs#L37

**Reply by Andrew15-5** [reply]
Why in the default session manager plugin which uses this function there is no request of this permission?

https://github.com/zellij-org/zellij/blob/56baf0b13dfc177a095ae9b3267fe531d4453099/default-plugins/session-manager/src/main.rs#L54-L68

After running the snippet from above (`zellij plugin -fs -- file:target/wasm32-wasi/debug/rust-plugin-example.wasm`) it compiles the plugin, and then it closes due to `close_self();`. Or does nothing if only `switch_session()` is called. The addition of the `request_permission(&[PermissionType::ChangeApplicationState]);` changes nothing.

```rs
    fn load(&mut self, _configuration: BTreeMap<String, String>) {
        request_permission(&[PermissionType::ChangeApplicationState]);
        switch_session(Some("existing-session"));
        close_self();
    }
```

Also, at some point, the command that starts the plugin can and did stop working by just running indefinitely without any output. So the development session is now only for writing code and a different session is for running the plugin. Probably some kind of bug that kills the plugin starting ability in the current session.

**Reply by cristiand391** [reply]
> Why in the default session manager plugin which uses this function there is no request of this permission?

interesting, haven't checked yet but maybe there's some internal logic that grant built-in plugins all required perms.

> The addition of the request_permission(&[PermissionType::ChangeApplicationState]); changes nothing.

On the first load it should prompt you to confirm access to the requested permissions, did you see that screen?
What I usually do is to nuke zellij's cache and reload the plugin.

Also keep in mind that plugin commands don't block until they are done:
https://github.com/zellij-org/zellij/issues/2558

it could be that `close_self` ends up running before `switch_session` is done 🤷🏼 

kinda related:
yesterday I wrote this plugin that uses the `kill_sessions` command:
https://github.com/zellij-org/zellij/issues/467#issuecomment-2134344111

something I noticed is that the UI wasn't printing the session name until ~1s after being rendered (which I get from the SessionUpdate event), so I had to make it wait until it gets it like this:
https://github.com/cristiand391/zj-quit/blob/c98e0741bd45d7c1de105e0055d5e0e504e76c65/src/main.rs#L43

you could try copying the logic from there to ensure `switch_session` only runs when the session info is available.

**Reply by Andrew15-5** [reply]
> On the first load it should prompt you to confirm access to the requested permissions, did you see that screen?

Yes, a few times.

> What I usually do is to nuke zellij's cache and reload the plugin.

How do you nuke tho cache? My only guess is that somehow I didn't grant the permission, and it just don't work now.

> it could be that `close_self` ends up running before `switch_session` is done

Damn, that's rough...

Can I use those functions in any of the 3 callbacks, or only in the `load`/`update`?

> you could try copying the logic from there to ensure `switch_session` only runs when the session info is available.

But why do I have to wait if I just need to switch to another session? Or I can't until that info is available? If that's true, I think the switch wouldn't be instant... :/

**Reply by cristiand391** [reply, marked as answer]
> How do you nuke tho cache? My only guess is that somehow I didn't grant the permission, and it just don't work now.

on linux I run `rm -rf $HOME/.cache/zellij`. I think on macos is somewhere in `~/Library/Cache/zellij`, I can confirm tomorrow.

> Can I use those functions in any of the 3 callbacks, or only in the load/update?

I'm pretty sure in any of the 4 lifecycle methods (4th is the pipe one, documented in other section):
https://zellij.dev/documentation/plugin-lifecycle

> But why do I have to wait if I just need to switch to another session? Or I can't until that info is available? If that's true, I think the switch wouldn't be instant... :/

Actually I don't think you would need to wait for `switch_session`. 
I ran your example above but without calilng `close_self`, and after giving perms I see this line in the log:

```
ERROR  |zellij_server::plugins::z| 2024-05-28 22:27:52.365 [async-std/runti] [zellij-server/src/plugins/zellij_exports.rs:273]: Plugin '/home/cristiand391/code/gh/zj-switch-session-test/target/wasm32-wasi/debug/rust-plugin-example.wasm (ID 4)' permission 'ChangeApplicationState' denied - Command 'SwitchSession' denied 
```

that reminded me of the `PermissionRequestResult` event doc'd here:
https://zellij.dev/documentation/plugin-lifecycle.html#update

basically the plugin was calling `switch_session` before it actually requested the perms because `request_permission` didn't finish yet!

following the linked blog I was able to make this work by only running `switch_session` when it's confirmed the permission was granted:

[Screencast from 05-28-2024 11:06:23 PM.webm](https://github.com/zellij-org/zellij/assets/6853656/b801e686-c492-48ab-b886-98ed3eeae280)

here's the code:
https://gist.github.com/cristiand391/3a40974034b2731ca5e866850ec599d1

**Reply by Andrew15-5** [reply]
> on linux I run `rm -rf $HOME/.cache/zellij`. I think on macos is somewhere in `~/Library/Cache/zellij`, I can confirm tomorrow.

Oh, so the normal way. Ok. (Linux FTW!)

I don't see any documentation of the `PermissionType::ReadApplicationState` that is needed for `EventType::PermissionRequestResult`, but looks like it's necessary...

Ok, so, I finally was able to do the plugin. But there are still 2 issues: I can't hide it because it needs permission the first time and sometimes when I launch it, it will still take about 0.5 or 1 second. But in some cases the switch is kinda instant.

I made switcher that will just switch and replacer, that will kill the initial session, because I usually call my alias from an empty terminal anyway. Hopefully I will have time later to document it and release.

Thanks for the help.
